README.md
libfidwid
=========

Fast Image Display library that supports drawing to X11 or OpenGL. It
was originally written for high performance camera image display in 2000.

### Installation

```bash
autoreconf -fiv
./configure
make
```

### Tests

The `src/testFidWid` contains four test programs:
- `testFidWidget` (draws using XPutImage, original version)
- `testOpenGLFBWidget` (draws to the OpenGL framebuffer)
- `testOpenGLTexWidget` (draws to an OpenGL texture)
- `testX11FidWidget` (draws using XPutImage, optimized)

### Benchmarks

On a i3-2357M CPU @ 1.30GHz with Intel HD 3000 graphics, the following
timing results were achieved (using the included `test.ppm`):

- `testFidWidget`: 116.6 FPS
- `testOpenGLFBWidget`: 53.8 FPS
- `testOpenGLTexWidget`: 161.8 FPS
- `testX11FidWidget`: 158.2 FPS


debian/changelog
libfidwid (0.1.0-1) unstable; urgency=low

  * Initial release

 -- Joerg Mechnich <mechnich@informatik.uni-freiburg.de>  Fri, 19 May 2006 22:36:10 +0200



src/testFidWid/testFidWid.cc
/**************************************************************************
**       Title: test FastImageDisplayWidget
**    $RCSfile: testFidWid.cc,v $
**   $Revision: 1.4 $$Name:  $
**       $Date: 2005/06/21 11:21:46 $
**   Copyright: GPL $Author: emmenlau $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: testFidWid.cc,v $
**  Revision 1.4  2005/06/21 11:21:46  emmenlau
**   - changed cast: new unsigned char[imagesize*3] (gcc-4.0 compliant)
**
**  Revision 1.3  2005/02/27 23:09:59  mechnich
**  re-added old X11 widget
**
**  Revision 1.2  2004/07/26 15:20:18  mechnich
**  added OpenGL display widgets and tests
**
**  Revision 1.1.1.1  2003/01/23 12:30:31  ronneber
**  initial revision
**
**  Revision 1.1  2002/01/10 17:50:17  ronneber
**  initial revision
**
**
**
**************************************************************************/

#include <ctime>
#include <iostream>
#include <fstream>
#include <iomanip>
using namespace std;

#include <qapplication.h>
#include <qobject.h>
#include <PixelConverter.hh>

#ifdef WIDGETTYPE
#include <FastImageDisplayWid_Templ.hh>
#include <X11FidWidget.hh>
#include <OpenGLFidWidget.hh>
typedef FastImageDisplayWid_Templ<WIDGETTYPE> FastImageDisplayWid;
#else
#include <FastImageDisplayWid.hh>
#endif

static std::istream& 
_eatWsAndComments(std::istream &s)
{
  s >> std::ws;
          
  while (s.peek()=='#')
  {
    s.ignore(10000, '\n');  // ignore characters until line end
    s >> std::ws;
  }
  return s;
}

struct EXAMPLE_RGB8
{
  unsigned char red, green, blue;
};


int main( int argc, char** argv)
{
  QApplication app( argc, argv);
  
  std::string filename = "test.ppm";
  std::cout << "Trying '" << filename << "'" << std::endl;
  std::ifstream infile( filename.c_str());
  if( !infile.is_open())
  {
    std::cout << "Could not open '" << filename << "'" << std::endl;
    filename = std::string(DATADIR) + "/" + filename;
    std::cout << "Trying '" << filename << "'" << std::endl;
    infile.open( filename.c_str());
    if( !infile.is_open())
    {
      std::cout << "Could not open '" << filename << "', exiting" << std::endl;
      std::exit(1);
    }
  }
  
  std::string magic;
  int width, height, maxvalue;
  infile >> magic 
         >> _eatWsAndComments >> width 
         >> _eatWsAndComments >> height
         >> _eatWsAndComments >> maxvalue;
  char dummyChar;
  infile.get(dummyChar);

  int imagesize =  width * height;
  unsigned char* image = new unsigned char[imagesize*3];
  infile.read( reinterpret_cast<char*>(image), imagesize*3);
 
  /*---------------------------------------------------------------------
   *  Create a pixel converter object
   *---------------------------------------------------------------------*/
  RGB_RGBConverter conv;
    
  /*---------------------------------------------------------------------
   *  Create a FastImageDisplayWidget
   *---------------------------------------------------------------------*/
  FastImageDisplayWid fidw;
  fidw.setImageSize( width, height);
  fidw.show();
 
  /*---------------------------------------------------------------------
   *  Run our own main loop for Qt
   *---------------------------------------------------------------------*/
  int frames = 0;

  clock_t start_time = clock();
  
  while( fidw.isVisible())
  {
    /*-------------------------------------------------------------------
     *  process image
     *-------------------------------------------------------------------*/
    EXAMPLE_RGB8* p = reinterpret_cast<EXAMPLE_RGB8*>( image);
    for( int i = 0; i < imagesize; ++i)
    {
      p->red -= 2;
      p->green -= 2;
      p->blue -= 2;
      ++p;
    }
    /*-------------------------------------------------------------------
     *  display image
     *-------------------------------------------------------------------*/
    fidw.setImagePerLine( image, conv);
      
    /*-------------------------------------------------------------------
     *  process Qt Events for at least 1 ms
     *-------------------------------------------------------------------*/
    app.processEvents();

    /*-------------------------------------------------------------------
     *  Display frames per second
     *-------------------------------------------------------------------*/
    ++frames;
    float elapsed_time = (float)(clock() - start_time) / CLOCKS_PER_SEC;
    if( elapsed_time > 1 )
    {
      cout << "frames per second = " << (float)frames / elapsed_time
           << endl;
      frames = 0;
      start_time = clock();
    }
  }
    
  return 0;
}


src/libfidwid/PixelConverter.cc
/**************************************************************************
**       Title: 
**    $RCSfile: PixelConverter.cc,v $
**   $Revision: 1.1 $$Name:  $
**       $Date: 2004/07/26 15:21:12 $
**   Copyright: GPL $Author: mechnich $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: PixelConverter.cc,v $
**  Revision 1.1  2004/07/26 15:21:12  mechnich
**  initial revision
**
**
**
**************************************************************************/

#include "PixelConverter.hh"

PixelConverter::_uint16
PixelConverter::_red16Table[256];
PixelConverter::_uint16
PixelConverter::_green16Table[256];
PixelConverter::_uint16
PixelConverter::_blue16Table[256];

PixelConverter::_uint16
PixelConverter::_convTable8To16[256];

bool
PixelConverter::_uninitialized = true;

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:
 *  ==> see headerfile
 *=======================================================================*/
PixelConverter::PixelConverter()
{
  /*-----------------------------------------------------------------------
   *  Initialize Conversion Tables
   *-----------------------------------------------------------------------*/
  if( _uninitialized)
  {
    for( unsigned int i=0; i < 256; ++i)
    {
      _red16Table[i] = (i & 0xf8) << 8;
      _green16Table[i] = (i & 0xfc) << 3;
      _blue16Table[i] = (i & 0xf8) >> 3;

      _convTable8To16[i]=i*256;
    }
    _uninitialized = false;
  }
}


src/libfidwid/OpenGLFidWidget.cc
/**************************************************************************
**       Title: 
**    $RCSfile: OpenGLFidWidget.cc,v $
**   $Revision: 1.2 $$Name:  $
**       $Date: 2005/07/14 09:33:32 $
**   Copyright: GPL $Author: tschmidt $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: OpenGLFidWidget.cc,v $
**  Revision 1.2  2005/07/14 09:33:32  tschmidt
**  - added SubSampling capabilities
**
**  Revision 1.1  2004/07/26 15:21:12  mechnich
**  initial revision
**
**
**
**************************************************************************/

#include "OpenGLFidWidget.hh"


/*-------------------------------------------------------------------------
 *  OpenGLFBWidget class
 *-------------------------------------------------------------------------*/

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  Constructor
 *  ==> see headerfile
 *=======================================================================*/
OpenGLFBWidget::OpenGLFBWidget( QWidget * parent)
        :QGLWidget( parent), _subSampling(1), _data( 0),
         _rowStart( 0)
{}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  Destructor
 *  ==> see headerfile
 *=======================================================================*/
OpenGLFBWidget::~OpenGLFBWidget()
{
  if( _data)
      delete _data;
  if( _rowStart)
      delete[] _rowStart;
}


/*-------------------------------------------------------------------------
 *  OpenGLTexWidget class
 *-------------------------------------------------------------------------*/

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  Constructor
 *  ==> see headerfile
 *=======================================================================*/
OpenGLTexWidget::OpenGLTexWidget( QWidget * parent)
        :QGLWidget( parent), _subSampling(1), _data( 0),
         _rowStart( 0), _angle( 0),
         _tex_width( 0), _tex_height( 0)
{
  for( unsigned int i=0; i < 16; ++i)
      _texMat[i] = 0.0;

  _texMat[10] = _texMat[15] = 1.0;
}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION: Destructor
 *  ==> see headerfile
 *=======================================================================*/
OpenGLTexWidget::~OpenGLTexWidget()
{
  if( _data)
      delete _data;
  if( _rowStart)
      delete[] _rowStart;
}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:
 *  ==> see headerfile
 *=======================================================================*/
void
OpenGLTexWidget::keyPressEvent( QKeyEvent* e)
{
  switch( e->key())
  {
  case Qt::Key_Left:
    _angle -=1;
    _angle %= 360;
    e->accept();
    break;
  case Qt::Key_Right:
    _angle +=1;
    _angle %= 360;
    e->accept();
    break;
  case Qt::Key_Q:
  case Qt::Key_Escape:
    close();
    e->accept();
    break;
  default:
    break;
  }
}


src/libfidwid/FastImageDisplayWid.cc
/**************************************************************************
**       Title: Widget for fast display of images 
**    $RCSfile: FastImageDisplayWid.cc,v $
**   $Revision: 1.4 $$Name:  $
**       $Date: 2005/07/14 09:26:30 $
**   Copyright: DWD GF-MM $Author: tschmidt $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: FastImageDisplayWid.cc,v $
**  Revision 1.4  2005/07/14 09:26:30  tschmidt
**  - added SubSampling capabilities
**
**  Revision 1.3  2005/02/27 23:12:05  mechnich
**  re-added old X11 widget
**
**  Revision 1.2  2003/10/23 17:21:45  mechnich
**  added depth() and data() functions
**
**  Revision 1.1.1.1  2003/01/23 12:30:31  ronneber
**  initial revision
**
**  Revision 1.1  2002/01/10 17:52:16  ronneber
**  imported sources from last year
**
**
**************************************************************************/

#include "FastImageDisplayWid.hh"

#include <iostream>
#include <QX11Info>

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  Constructor
 *  ==> see headerfile
 *=======================================================================*/
FastImageDisplayWid::FastImageDisplayWid( QWidget * parent)
        :QWidget( parent),
         _xImage( 0), _depth( QX11Info::appDepth()), _subSampling(1)
{
  setAttribute( Qt::WA_NoBackground);
  setAttribute( Qt::WA_PaintOnScreen);
  /*-----------------------------------------------------------------------
   *  create the GC
   *-----------------------------------------------------------------------*/
  _imageGC = XCreateGC(QX11Info::display(), winId(), 0, NULL);
}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  setImageSize()
 *  ==> see headerfile
 *=======================================================================*/
void FastImageDisplayWid::setImageSize( unsigned int width, unsigned int height)
{

  unsigned int subWidth = width / _subSampling;
  unsigned int subHeight = height / _subSampling;

  /*-----------------------------------------------------------------------
   *  Check if xImage needs to be resized
   *-----------------------------------------------------------------------*/
  if( _xImage == 0 
      || subHeight != (unsigned int)_xImage->height 
      || subWidth != (unsigned int)_xImage->width)
  {
    if( _xImage != 0)
    {
      XDestroyImage( _xImage);
    }

    /*---------------------------------------------------------------------
     *  find out, how man bytes per pixel are required
     *---------------------------------------------------------------------*/
    int padding;
    
    switch( _depth)
    {
    case 16:
      std::cerr << "Bit depth not supperted !" << std::endl;
      padding = 2;
      break;
    case 24:
      padding = 4;
      break;
    case 32:
      padding = 4;
      break;
    default:
    {
      std::ostringstream errmsg;
      errmsg << "Requested bit depth (" << _depth << ") not supported. "
          "Please Set your Screen to 16, 24 or 32 bits, or use command "
          "line argument '-visual TrueColor' on SGI's";
      throw UnsupportedBitDepth( errmsg.str());
    }
    }
    
    /*---------------------------------------------------------------------
     *  create an XImage (ownership of allocated data is passed to xImage)
     *---------------------------------------------------------------------*/
    _xImage = XCreateImage(QX11Info::display(), 
                           (Visual*)QX11Info::appVisual(),
                           QX11Info::appDepth(), 
                           ZPixmap, 0,
                           (char*) malloc( subHeight * subWidth * padding), 
                           subWidth, subHeight, padding*8, 0);
  }

  /*-----------------------------------------------------------------------
   *  the QtWidget should not be resized to bigger sizes
   *-----------------------------------------------------------------------*/
  setFixedSize( subWidth, subHeight);
}


/*=========================================================================
 *  DESCRIPTION OF FUNCTION:
 *  ==> see headerfile
 *=======================================================================*/
QSize FastImageDisplayWid::sizeHint() const
{
  return QSize( _xImage->width, _xImage->height);
}



/*=========================================================================
 *  DESCRIPTION OF FUNCTION:
 *  ==> see headerfile
 *=======================================================================*/
void FastImageDisplayWid::paintEvent( QPaintEvent * )
{
  /*-----------------------------------------------------------------------
   *  Just redraw the whole widget (don't take regions into accout)
   *-----------------------------------------------------------------------*/
  if( _xImage != 0)
  {
    XPutImage(QX11Info::display(), winId(), _imageGC, _xImage,
              0, 0, 0, 0, _xImage->width, _xImage->height);
  }
}



src/libfidwid/X11FidWidget.cc
/**************************************************************************_
**       Title: 
**    $RCSfile: X11FidWidget.cc,v $
**   $Revision: 1.3 $$Name:  $
**       $Date: 2005/07/14 09:35:22 $
**   Copyright: GPL $Author: tschmidt $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: X11FidWidget.cc,v $
**  Revision 1.3  2005/07/14 09:35:22  tschmidt
**  - added SubSampling capabilities
**
**  Revision 1.2  2005/06/21 11:21:21  emmenlau
**   - changed cast: new unsigned char*[height] (gcc-4.0 compliant)
**
**  Revision 1.1  2004/07/26 15:21:12  mechnich
**  initial revision
**
**
**
**************************************************************************/

#include "X11FidWidget.hh"

#include <iostream>

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  Constructor
 *  ==> see headerfile
 *=======================================================================*/
X11FidWidget::X11FidWidget( QWidget * parent)
:QWidget( parent),
    _xImage( 0), _depth( QX11Info::appDepth()), _rowStart( 0), _subSampling(1)
{
  setAttribute( Qt::WA_NoBackground);
  setAttribute( Qt::WA_PaintOnScreen);
  /*-----------------------------------------------------------------------
   *  create the GC
   *-----------------------------------------------------------------------*/
  _imageGC = XCreateGC(QX11Info::display(), winId(), 0, NULL);
}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  Destructor
 *  ==> see headerfile
 *=======================================================================*/
X11FidWidget::~X11FidWidget()
{
  if( _xImage)
      XDestroyImage( _xImage);
  
  if( _rowStart)
      delete[] _rowStart;
}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:  setImageSize()
 *  ==> see headerfile
 *=======================================================================*/
void
X11FidWidget::setImageSize( unsigned int width, unsigned int height)
{

  unsigned int subWidth = width / _subSampling;
  unsigned int subHeight = height / _subSampling;

//   std::cerr << "X11FidWidget::setImageSize(): " << std::endl
//             << "  original size:   " << width << "x" << height << std::endl
//             << "  subSampled size: " << subWidth << "x" << subHeight << std::endl;
  
  /*-----------------------------------------------------------------------
   *  Check if xImage needs to be resized
   *-----------------------------------------------------------------------*/
  if( _xImage == 0 
      || subHeight != (unsigned int)_xImage->height 
      || subWidth != (unsigned int)_xImage->width)
  {
    if( _xImage)
    {
      XDestroyImage( _xImage);
    }

    /*---------------------------------------------------------------------
     *  find out, how man bytes per pixel are required
     *---------------------------------------------------------------------*/
    int padding=0;
    
    switch( _depth)
    {
    case 16:
      padding = 2;
      break;
    case 24:
      padding = 4;
    case 32:
      padding = 4;
      break;
    default:
      std::cerr << "Unsupported display depth " << depth() << std::endl;
      exit(1);
      break;
    }
    
    int lineWidth = subWidth * padding;
    /*---------------------------------------------------------------------
     *  create an XImage (ownership of allocated data is passed to xImage)
     *---------------------------------------------------------------------*/
    _xImage = XCreateImage( QX11Info::display(), 
                            (Visual*)QX11Info::appVisual(),
                            _depth, 
                            ZPixmap, 0,
                            (char*) malloc( subHeight * lineWidth), 
                            subWidth, subHeight, padding*8, 0);

//    std::cerr << "X11FidWidget::setImageSize(): new XImage created" << std::endl;
    
    if( _rowStart)
        delete[] _rowStart;
    _rowStart = new unsigned char*[subHeight];
    for( unsigned i=0; i < subHeight; ++i)
    {
      _rowStart[i] = reinterpret_cast<unsigned char*>
          (_xImage->data + i * lineWidth);
    }

//    std::cerr << "X11FidWidget::setImageSize(): lineStartPointers set" << std::endl;

    setFixedSize(subWidth, subHeight);
    
  }
}

/*=========================================================================
 *  DESCRIPTION OF FUNCTION:
 *  ==> see headerfile
 *=======================================================================*/
void
X11FidWidget::paintEvent( QPaintEvent* repaintEvent)
{
  /*-----------------------------------------------------------------------
   *  Redraw the rectangular portion of the Image the QPaintEvent requests
   *-----------------------------------------------------------------------*/
  if( _xImage != 0)
  {
    
//     std::cerr << "X11FidWidget::paintEvent(): do repaint... " << std::flush;
    
    
    XPutImage(QX11Info::display(), winId(), _imageGC, _xImage,
              repaintEvent->rect().left(), repaintEvent->rect().top(),
              repaintEvent->rect().left(), repaintEvent->rect().top(),
              repaintEvent->rect().width(), repaintEvent->rect().height());

//     std::cerr << "done" << std::endl;

  }
}
